2016/presentations/Debugging_of_CPython_processes_with_gdb/text.md

```diff
@@ -0,0 +1,639 @@
+runtime (especially with aggressive optimizations enabled by `-O3`). In this
+case `gdb` will show you something like:
+
+```
+
+(gdb) bt
+
+#0  0x00007fdf3ca31be3 in __select_nocancel () at ../sysdeps/unix/syscall-template.S:84
+#1  0x00000000005d1da4 in pysleep (secs=<optimized out>) at ../Modules/timemodule.c:1408
+#2  time_sleep () at ../Modules/timemodule.c:231
+#3  0x00000000004f5465 in call_function (oparg=<optimized out>, pp_stack=0x7fff62b184c0) at ../Python/ceval.c:4637
+#4  PyEval_EvalFrameEx () at ../Python/ceval.c:3185
+#5  0x00000000004f5194 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fff62b185c0, 
+    func=<optimized out>) at ../Python/ceval.c:4750
+#6  call_function (oparg=<optimized out>, pp_stack=0x7fff62b185c0) at ../Python/ceval.c:4677
+#7  PyEval_EvalFrameEx () at ../Python/ceval.c:3185
+#8  0x00000000004f5194 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fff62b186c0, 
+    func=<optimized out>) at ../Python/ceval.c:4750
+#9  call_function (oparg=<optimized out>, pp_stack=0x7fff62b186c0) at ../Python/ceval.c:4677
+#10 PyEval_EvalFrameEx () at ../Python/ceval.c:3185
+#11 0x00000000005c5da8 in _PyEval_EvalCodeWithName.lto_priv.1326 () at ../Python/ceval.c:3965
+#12 0x00000000005e9d7f in PyEval_EvalCodeEx () at ../Python/ceval.c:3986
+#13 PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at ../Python/ceval.c:777
+#14 0x00000000005fe3d2 in run_mod () at ../Python/pythonrun.c:970
+#15 0x000000000060057a in PyRun_FileExFlags () at ../Python/pythonrun.c:923
+#16 0x000000000060075c in PyRun_SimpleFileExFlags () at ../Python/pythonrun.c:396
+#17 0x000000000062b870 in run_file (p_cf=0x7fff62b18920, filename=0x1733260 L"test2.py", fp=0x1790190) at ../Modules/main.c:318
+#18 Py_Main () at ../Modules/main.c:768
+#19 0x00000000004cb8ef in main () at ../Programs/python.c:69
+#20 0x00007fdf3c970610 in __libc_start_main (main=0x4cb810 <main>, argc=2, argv=0x7fff62b18b38, init=<optimized out>, fini=<optimized out>, 
+    rtld_fini=<optimized out>, stack_end=0x7fff62b18b28) at libc-start.c:291
+#21 0x00000000005c9df9 in _start ()
+
+(gdb) py-bt
+Traceback (most recent call first):
+  File "test2.py", line 9, in g
+    time.sleep(1000)
+  File "test2.py", line 5, in f
+    g()
+  (frame information optimized out)
+
+```
+
+i.e. some application level frames will be available, some will not.
+There is little you can do at this point, except for rebuilding CPython
+with a lower optimization level, but that often is not an option for production
+(not to mention the fact you'll be using a custom CPython build, not the
+one provided by your Linux distro).
+
+[optimization level]: https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html
+
+
+## Virtual environments and custom CPython builds
+
+When a virtual environment is used, it may appear that the extension does not work:
+
```

Comment:
It would be great to add the command used to run gdb here: **gdb -p 2975** - I found it a bit confusing to not know (at this point, as command can be found later) how it was run with virtualenv case (with or without path to executable?)